jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java

```diff
@@ -18,57 +18,121 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import org.eclipse.jetty.client.api.Connection;
 import org.eclipse.jetty.client.api.Destination;
-import org.eclipse.jetty.util.AtomicBiInteger;
+import org.eclipse.jetty.util.Attachable;
 import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.Pool;
 import org.eclipse.jetty.util.Promise;
 import org.eclipse.jetty.util.annotation.ManagedAttribute;
 import org.eclipse.jetty.util.annotation.ManagedObject;
 import org.eclipse.jetty.util.component.Dumpable;
 import org.eclipse.jetty.util.log.Log;
 import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.thread.Sweeper;
+
+import static java.util.stream.Collectors.toList;
 
 @ManagedObject
-public abstract class AbstractConnectionPool implements ConnectionPool, Dumpable
+public abstract class AbstractConnectionPool implements ConnectionPool, Dumpable, Sweeper.Sweepable
 {
     private static final Logger LOG = Log.getLogger(AbstractConnectionPool.class);
 
-    /**
-     * The connectionCount encodes both the total connections plus the pending connection counts, so both can be atomically changed.
-     * The bottom 32 bits represent the total connections and the top 32 bits represent the pending connections.
-     */
-    private final AtomicBiInteger connections = new AtomicBiInteger();
-    private final AtomicBoolean closed = new AtomicBoolean();
     private final HttpDestination destination;
     private final int maxConnections;
     private final Callback requester;
+    private final AtomicInteger pendingConnectionCount = new AtomicInteger();
+    private final Pool<Connection> pool;
 
     /**
-     * @param destination the correspondent destination
-     * @param maxConnections the max number of connections
-     * @param requester the callback to notify about new connection creation/failure
-     * @deprecated use {@link #AbstractConnectionPool(HttpDestination, int, Callback)} instead
+     * @deprecated use {@link #AbstractConnectionPool(HttpDestination, int, boolean, Callback)} instead
      */
     @Deprecated
     protected AbstractConnectionPool(Destination destination, int maxConnections, Callback requester)
     {
-        this((HttpDestination)destination, maxConnections, requester);
+        this((HttpDestination)destination, maxConnections, true, requester);
     }
 
-    protected AbstractConnectionPool(HttpDestination destination, int maxConnections, Callback requester)
+    protected AbstractConnectionPool(HttpDestination destination, int maxConnections, boolean cache, Callback requester)
     {
         this.destination = destination;
         this.maxConnections = maxConnections;
         this.requester = requester;
+        this.pool = new Pool<>(cache ? 1 : 0);
+        destination.addBean(this.pool);
```

Comment:
If the pool is being added to the destination as a been, then perhaps we should first look for an existing pool on the destination before creation a new one?

Reply:
Sure, done.